### PR TITLE
Close clipped paths

### LINKED ISSUE #4185: Colorbar outline has broken path in vector backends
Hi again,
In all of the vector graphics backends that I tested (pdf, eps, ps, svg), the colorbar outline is drawn with a broken path at the corner where the path starts and ends.  This does not happen with the raster graphic or interactive backends.  Here is a minimal working example:
```python
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

mpl.rcParams.update({'axes.linewidth':10})

fig = plt.figure(figsize=(8,6))
ax1 = fig.add_axes([0.05, 0.85, 0.9, 0.1])
ax2 = fig.add_axes([0.05, 0.65, 0.9, 0.1])
ax3 = fig.add_axes([0.05, 0.45, 0.9, 0.1])
ax4 = fig.add_axes([0.05, 0.25, 0.9, 0.1])
ax5 = fig.add_axes([0.05, 0.05, 0.9, 0.1])

cmap = mpl.cm.jet
cmap.set_under('w')
cmap.set_over('w')

im = ax1.pcolormesh(np.linspace(0,10,16).reshape((4,4)))

plt.colorbar(im,cax=ax2,cmap=cmap,orientation='horizontal',
             extend='both',extendfrac=0.5)
plt.colorbar(im,cax=ax3,cmap=cmap,orientation='horizontal',
             extend='both',)
plt.colorbar(im,cax=ax4,cmap=cmap,orientation='horizontal',
             extend='both',extendrect=True)
plt.colorbar(im,cax=ax5,cmap=cmap,orientation='horizontal',
             extend='neither')

plt.savefig('colorbar_tip.pdf')
plt.savefig('colorbar_tip.eps')
plt.savefig('colorbar_tip.svg')
```
Tested with matplotlib 1.4.3, and python 2.7, on Linux and Windows. I have exaggerated the axes linewidth to make the problem more visible.  Here is a screenshot of the PDF output:

colorbar_tip

I have been able to fix this by making the following change to the `ColorbarBase` class in colorbar.py:
```python
def _config_axes(self, X, Y):

    ...

    if self.outline is not None:
        self.outline.remove()

    xy2 = np.append(xy,xy[1:2,:],axis=0) # <-- added

    self.outline = mpatches.Polygon(
        xy2, edgecolor=mpl.rcParams['axes.edgecolor'], # <-- modified
        facecolor='none',
        linewidth=mpl.rcParams['axes.linewidth'],
        closed=True,
        zorder=2)

    ...
```
This essentially just copies the second point of the path to the end of the path.  However, I don't know if this is the best fix.  I don't know anything about vector graphics drawing directives, but it seems to me that there should be a builtin function to have paths wrap-around enough so that it creates a closed loop, including drawing the correct corner.  Clearly the `closed=True` option of the `mpatches.Polygon` is not doing this.

Again, thanks to anyone who looks into this.

### PR DESCRIPTION
**Pull request by mdboom**
Fix #4185.

### CODE REVIEW

**Comment by WeatherGod** [comment]
one of the failures has to do with the recent fix to trifinder. Did numpy
1.6 not have "order" kwarg for its copy function? Also, didn't that PR pass
all of the tests?

On Tue, Mar 3, 2015 at 7:00 PM, Michael Droettboom <notifications@github.com

File: lib/matplotlib/tests/test_colorbar.py [hunk 1]
```diff
@@ -253,6 +254,33 @@ def test_colorbarbase():
...
+
+    im = ax1.pcolormesh(np.linspace(0,10,16).reshape((4,4)))
+
+    plt.colorbar(im,cax=ax2,cmap=cmap,orientation='horizontal',
+                 extend='both',extendfrac=0.5)
+    plt.colorbar(im,cax=ax3,cmap=cmap,orientation='horizontal',
+                 extend='both',)
+    plt.colorbar(im,cax=ax4,cmap=cmap,orientation='horizontal',
+                 extend='both',extendrect=True)
+    plt.colorbar(im,cax=ax5,cmap=cmap,orientation='horizontal',
```
Comment (WeatherGod) [inline lib/matplotlib/tests/test_colorbar.py]: this test is failing PEP8 because of the lack of spaces around the commas.

**Comment by tacaswell** [comment]
Would it be worth making the line width on the tests very large (given that we don't have _exact_ tests).

File: lib/matplotlib/tests/test_colorbar.py [hunk 2]
```diff
@@ -253,6 +254,33 @@ def test_colorbarbase():
...
+    remove_text=True)
+def test_colorbar_closed_patch():
+    fig = plt.figure(figsize=(8,6))
+    ax1 = fig.add_axes([0.05, 0.85, 0.9, 0.1])
+    ax2 = fig.add_axes([0.05, 0.65, 0.9, 0.1])
+    ax3 = fig.add_axes([0.05, 0.45, 0.9, 0.1])
+    ax4 = fig.add_axes([0.05, 0.25, 0.9, 0.1])
+    ax5 = fig.add_axes([0.05, 0.05, 0.9, 0.1])
+
+    cmap = cm.jet
```
Comment (efiring) [inline lib/matplotlib/tests/test_colorbar.py]: Maybe the image files would be smaller if the map had fewer colors?

Comment (tacaswell) [inline lib/matplotlib/tests/test_colorbar.py]: Also in the spirit of doing-in-jet should we use a different color map ;)

Comment (u55) [inline lib/matplotlib/tests/test_colorbar.py]: How about changing the cmap to this?:
```python
...
from matplotlib.colors import ListedColormap
cmap = ListedColormap([(.25,.25,.25,1.0),(.5,.5,.5,1.0),(.75,.75,.75,1.0)])
cmap.set_under('w')
cmap.set_over('w')

im = ax1.pcolormesh(np.linspace(0,10,16).reshape((4,4)),cmap=cmap)
...
```

**Comment by mdboom** [comment]
I think everything above is now addressed.

**Comment by tacaswell** [comment]
The clipping is different for the first on the pdf result and the svg/png result.
so

**Comment by mdboom** [comment]
@tacaswell: Yes, the default join style appears to be different (miter vs. bevel).  I'd rather treat that as a separate issue, as that's likely to affect many more test cases.

**Comment by u55** [comment]
@tacaswell @mdboom 

> The clipping is different for the first on the pdf result and the svg/png result.

The clipping of the svg image is dependent on the viewer. On my windows machine, they look different in Chrome versus Firefox.

chrome_svg
firefox_svg

**Comment by u55** [comment]
Also, in inkscape, the svg image looks the same as in Firefox.  So Chrome is the oddball.

**Comment by mdboom** [comment]
The difference we are seeing is that the stroke width and angle are high enough that the miter limit is going into effect.  Here's a good explanation of miter limit:

http://tutorials.jenkov.com/svg/stroke.html#stroke-miterlimit

The default miter limit is somewhat dependent on the renderer (it's specified in the SVG standard, for example, but I'm not certain every implementation follows it), and what we should be doing is setting it explicitly to get uniform behavior.

Ideally, we'd add an API for the user to set the miter limit explicitly, but we've never had that before (i.e. it's not a regression to *not* have it), and doing so is sort of beyond the scope of this PR.  Instead, I think for now we just set it to some reasonable maximum (effectively disabling it), since if the user wants the miters to be limited, they can always set the join style to bevel anyway (not exactly the same thing, of course, but good enough for now).

Now that I have a better understanding of what's going on here, I'll probably just add the fix to this PR (since the test here serves as a good test for the other).  Stay tuned...

**Comment by mdboom** [comment]
Ok.  The miter limit change is now included.  It did not affect any existing tests, other than the new one in this PR.

**Comment by mdboom** [comment]
I've updated the test images so the miter points don't extend beyond the edge of the image.

**Comment by efiring** [comment]
@mdboom,  thank you for taking care of this so quickly.  It looks like it needs to be cherry-picked to the color_overhaul branch, correct?  I would be more comfortable with you or @tacaswell doing that; I'm shaky with that git operation.

**Comment by u55** [comment]
@mdboom , thank you.  I installed master on my Linux machine and your fix works great.  Again thanks for the fast work.

**Comment by tacaswell** [comment]
@mdboom This touches code that was greatly changed as part of the cxx refactoring.  Can you have a look at back-porting this?

**Pull request merged**